### DISCUSSION TITLE
Cannot install 3.2.2 on M1 mac 13.4 ventura

### DISCUSSION THREAD

**Discussion by vguerci** [rbenv/ruby-build#2229] [Build failures] [answered]
## Problem
Running `ruby-build -vk 3.2.2 .` is failing with:
```
compiling vm_trace.c
compiling yjit.c
make: /opt/homebrew/bin/gmkdir: Permission denied
make: *** [coroutine/arm64/.time] Error 1
make: *** Waiting for unfinished jobs....

BUILD FAILED (macOS 13.4 using ruby-build 20230717)
```

## Investigations
Initially faced that issue using `rbenv` or `asdf`.
Searched for similar issues here, and proposed workarounds w/o figuring out a proper solution.
Running `make install` from the build folder will bypass the error, but then install ruby w/o openssl.

I asked some coworkers to test, and two other people reported the same issue, making me think this is unlikely to be an environmental issue.

EDIT: apparently affects only M1 machines

**Comment by mislav** [comment]
I'm suspecting that `gmkdir` is provided by `coreutils` Homebrew formula. Could you temporarily `brew unlink coreutils` and try again?

**Reply by vguerci** [reply]
You are suspecting right. FYI `coreutils` is a dependency of `asdf` homebrew package.

Running after unlinking it logs a different error:
```
Last 10 log lines:
rbconfig.rb updated
generating encdb.h
generating enc.mk
make: ./miniruby: Permission denied
creating verconf.h
building rb_mjit_header.h
make: *** [encdb.h] Error 1
make: *** Waiting for unfinished jobs....
verconf.h updated
rb_mjit_header.h updated
```

**Reply by mislav** [reply]
@vguerci Is anything from this thread illuminating? https://github.com/rbenv/ruby-build/issues/992#issuecomment-247342449

**Reply by vguerci** [reply]
Sadly not, I noticed this thread already, while very similar, it just led me to `make install` which lacks OpenSSL

**Reply by mislav** [reply]
I'm out of ideas. BTW, Ruby 3.2.2 compiles just fine on my (relatively pristine) macOS 13.4.

You can try uploading the full build log as a gist link to this thread please

**Reply by vguerci** [reply, marked as answer]
I think I've just found the root cause, `RUBY_MAKE_OPTS="-j 1" ruby-build -vk 3.2.2 .` builds fine...

~We are mostly iOS Software engineers using Xcode version `14.3` / latest `14.3.1`.~
~I suspect this issue comes from specific Xcode versions, where ruby-build default `make -j 10` induce ***_something_***...~

EDIT: XCode is most likely out of cause, a third-party party tool like anti-virus or similar is suspected.

That explains why people with similar issues in #992  were managing to finish install manually with `make` (without `-j x`)

**Reply by vguerci** [reply]
Tried with various Xcode versions `14.1`, `14.3`, `14.3.1`, `15.0.0 Beta 5`, all showing similar behaviour (but sometimes different errors)
Maybe the issue is at OS level? `13.4 (22F66)`
This is puzzling me 🤔🧩

**Reply by vguerci** [reply]
Random idea, maybe farfetched, but we have a company policy that enforces us to have an anti-virus (Microsoft defender, did exclude ruby-build build folders from it w/o any luck) and also FileVault (filesystem encryption)
💡  Will test on a personal mac M1 with same Xcode / macOS versions...

**Reply by vguerci** [reply]
Maybe not that farfetched, build does not fail on my personal mac mini M1 (8 cores) with Xcode `14.3.1` and macOS `13.4.1`

**Reply by mislav** [reply]
I have FileVault enabled on my macOS 13.4.1 and it has never caused problems. 
```
$ `xcode-select --print-path`/usr/bin/gcc --version       
Apple clang version 14.0.3 (clang-1403.0.22.14.1)
Target: arm64-apple-darwin22.5.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

An antivirus, however, could be the culprit, although you already said you've tried to rule that out. Other than that, sometimes newer Xcode versions introduce incompatibilities with the Ruby build system, although newer Ruby versions try to iron those out as fast as their core team can catch it.

You are welcome to post a full build log as a gist here, but I think we're already pretty sure that ruby-build itself doesn't cause this issue, but that it's a general issue with Ruby being built under your specific environment, and so you could try to report it (with as much information as possible) to the Ruby issue tracker.

Thanks for all the info shared so far and I'm sorry this is such a bother!

**Reply by vguerci** [reply]
Using the same `gcc` version:
```sh
$ xcode-select --print-path`/usr/bin/gcc --version
Apple clang version 14.0.3 (clang-1403.0.22.14.1)
Target: arm64-apple-darwin22.5.0
Thread model: posix
InstalledDir: /Applications/Xcode-14.3.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

Yes, I agree `ruby-build` definitely out of cause, I don't think posting a full log will bring any value.
I already gave a look at Ruby issue tracker, and couldn't find a similar report.
Anyway, as the anti-virus, or maybe also macOS `parentalcontrolsd` which our enrollment tool Jamf uses is most likely causing this, I doubt the Ruby team can do anything. I did ask my IT department if they could disable these temporarily on my workstation to confirm these assumptions, waiting for them...

Thanks for your support and commitment, much appreciated :)

**Reply by vguerci** [reply]
"Fun" fact, I have similar permissions denied errors while trying to build erlang...
And ofc, `ASDF_CONCURRENCY=1` fixes that...

I hope this information will be useful to other people.
Still waiting for my IT department to diagnose which third-party tool is doing that.

**Comment by mislav** [comment]
The invocation `ruby-build -vk 3.2.2 .` looks like you're trying to install into the current directory. If the current directory isn't empty or has insufficient permissions, it might not work. Could you pass something else as argument, for example `ruby-build -vk 3.2.2 /tmp/ruby-test`? (It's fine if the directory does not exist— it will be created.)

**Reply by vguerci** [reply]
I'm in an empty folder created to reproduce this, but will do as you suggest, I guess i have to `rm -rf /tmp/ruby-test` before each re-attempt?
